README.md

```diff
@@ -58,32 +123,300 @@ client = Redis.new(semian: {
+
+There are two configuration values. It's generally quite problematic to choose
+good values and we're still experimenting with ways to figure out good ticket
+numbers.  Generally something below half the number of workers on the server for
+endpoints that are queried frequently has worked out well for us.
+
+* **tickets**. Number of workers that can concurrently access a resource.
+* **timeout**. Time to wait to acquire a ticket if there are no tickets left.
+  We recommend this to be `0` unless you have very few workers running (i.e.
+  less than ~5).
+
+## Defense line
+
+The finished defense line for resource access with circuit breakers and
+bulkheads then looks like this:
+
+![](http://cdn.shopify.com/s/files/1/0070/7032/files/image02_grande.png?8570)
+
+The RPC first checks the circuit; if the circuit is open it will raise the
+exception straight away which will trigger the fallback (the default fallback is
+a 500 response). Otherwise, it will try Semian which fails instantly if too many
+workers are already querying the resource. Finally the driver will query the
+data store. If the data store succeeds, the driver will return the data back to
+the RPC. If the data store is slow or fails, this is our last line of defense
+against a misbehaving resource. The driver will raise an exception after trying
+to connect with a timeout or after an immediate failure. These driver actions
+will affect the circuit and Semian, which can make future calls fail faster.
+
+## Failing gracefully
+
+Ok, great, we've got a way to fail fast with slow resources, how does that make
+my application more resilient? 
+
+Failing fast is only half the battle. It's up to you what you do with these
+errors, in the [session example](#real-world-example) we handle it gracefully by
+signing people out and disabling all session related functionality till the data
+store is back online. However, not rescuing the exception and simply sending
+`HTTP 500` back to the client faster will help with [capacity
+loss](#capacity-loss).
+
+### Exceptions inherit from base class
+
+It's important to understand that the exceptions raised by [Semian
+Adapters](#adapters) inherit from the base class of the driver itself, meaning
+that if you do something like:
+
+```ruby
+def posts
+  Post.all
+rescue Mysql2::Error
+  []
+end
+```
+
+Exceptions raised by Semian's `MySQL2` adapter will also get caught.
+
+### Patterns
+
+However, we do not recommend mindlessly sprinkling `rescue`s all over the place.
+We recommend writing decorators around secondary data stores (e.g. sessions)
+that provide resiliency for free. For example, if we stored the tags associated
+with products in a secondary data store a subset of it could look like this:
+
+```ruby
+module ProductTags
+  def tags
+    tags_redis.get("product:tags:#{self.id}")
+  rescue Redis::CannotConnectError
+    []
+  end
+
+  private
+  def tags_redis
+    @redis ||= Redis.new
+  end
+end
+
+class Product
+  include ProductTags
```

Comment:
when you say "decorator", I'm assuming you are talking about the decorator design pattern? I'm no design pattern expert, but this thing here (using modules) doesn't look like that to me...?